design-proposals/vsock/vsock.md

```diff
@@ -0,0 +1,92 @@
+*   As a vendor I want to have an easy-to-use extension mechanism based on normal socket communication to easily develop reliable communication patterns with e.g. gRPC to maximize development speed while at the same time avoid vulnerabilities, bugs, …
+*   As a vendor, I want the ability to have intercommunication between a guest and a pod on the same host without having to use a network interface.
+
+### Background
+
+KubeVirt uses for local guest-host communication right now `virtio-serial`. This is currently used in KubeVirt by `libvirt` and `qemu` to communicate with the `qemu-guest-agent`. `virtio-serial` can also be used by other agents, but it is a little bit cumbersome due to:
+
+*   A small set of ports on the `virtio-serial` device
+*   Low bandwidth
+*   No socket based communication possible, which requires every agent to establish their own protocols, or work with translation layers like SLIP to be able to use protocols like gRPC for reliable communication.
+*   No easy and supportable way to get a `virtio-serial` socket assigned and being able to access it without entering the virt-launcher pod.
+*   Due to the point above, privileges are required for services.
+
+With `virtio-vsock` (https://man7.org/linux/man-pages/man7/vsock.7.html) we get support for easy guest-host communication which solves the above issues from a user/admin perspective.
+
+### Design
+
+The feature can be added to KubeVirt in two phases.
+
+*   Phase 1: node-local bits to wire in vsock, only allowing privileged node-local services to communicate with the Guest VMs.
+*   Phase 2: Add websocket proxying to virt-handler, to allow accessing VSOCK ports via REST calls and  protecting it via k8s RBAC.
+
+### Phase 1: node-local wiring of vsock
+
+*   The `/dev/vhost-vsock` device will be exposed via virt-handler under the `vsock` feature gate
+*   Each virt-launcher with `vsock` enabled will request the device
+*   Virt-controller will hand out a CID to virt-launcher and add it to the VMI status. 
+*   virt-launcher/libvirt/qemu take the CID and configure the vsock device via `/dev/vhost-vsock`
+*   From this point on `/dev/vhost-vsock` does not have any function anymore
+*   To simplify the migration story, the CID will be unique cluster wide. Therefore the CID will be retained after migration. (With 100 000 k8s nodes we would still have enough CIDs to run more than 20 000 VMs per node). The CID has a type of `uint32`, which is large enough to satisfy a cluster-wide unique allocation)
+
+![vsock flow](vsock.png)
+
+#### API changes
+
+A `autoattachVsock` flag will be added to the VMI spec at `spec.domain.devices` which defaults to `false` to stay backward compatible. Also a `VsockCID` field will be added to the VMI status as `status.vsockCID` for CID allocation and lookup.
```

Comment:
is there any chance that we'll need more fields associated with vsock other that "true|false" for attachment?  if so, maybe we shouldn't use a boolean.

Reply:
libvirt has three fields, one for auto-cid selection, one for the virtio model and one for manual cid assignment. We could not come up with additional use-cases where more fields would have to be on the VMI. For e.g. secure computing I would expect that vsock as a whole has to be disabled, but that's about it ...

Reply:
Can you think about something?

Reply:
Could that be that we will need more than one vsock per guest at some point?

Reply:
The vsock device has ports like TCP/IP. We can have a lot of agents over the single device. That is probably what you read about.